x-pack/legacy/plugins/snapshot_restore/public/app/app.tsx

```diff
@@ -21,16 +22,37 @@ export const App: React.FunctionComponent = () => {
     },
   } = useAppDependencies();
 
+  // Get app state to set permissions data
+  const [, dispatch] = useAppState();
+
+  // Use ref for default permission data so that re-rendering doesn't
+  // cause dispatch to be called over and over
+  const defaultPermissionsData = useRef({
+    hasPermission: true,
+    missingClusterPrivileges: [],
+    missingIndexPrivileges: [],
+  });
+
   // Load permissions
   const {
     error: permissionsError,
     loading: loadingPermissions,
-    data: { hasPermission, missingClusterPrivileges } = {
-      hasPermission: true,
-      missingClusterPrivileges: [],
-    },
+    data: permissionsData = defaultPermissionsData.current,
   } = useLoadPermissions();
 
+  const { hasPermission, missingClusterPrivileges } = permissionsData;
+
+  // Update app state with permissions data
+  useEffect(
+    () => {
+      dispatch({
```

Comment:
@elastic/es-ui 
I think permissions is a good candidate for the type of data that should be stored in React context. Here, the permissions are fetched when the app mounts and I dispatch the resulting data into the app state so that [other components can access it](https://github.com/elastic/kibana/pull/39966/files#diff-db8e4e06e339e3d5e49617830b0e85dcR47) without having to send another request to get permissions info. This can let us get really granular with the way the UI behaves at different permission levels. I've only used it in one section of this app for now, but it would be relatively easy to check for more granular permissions and disable buttons such as Delete and Restore.

Reply:
Agreed @jen-huang. And should be a reusable solution across our apps.  Once the Provider wraps the app, I see something like this in our components

```js
import { Authorize } from 'plugins/our_shared_es_ui_plugin/permissions';

const SomeComponent = () => (
  <Authorize role="someRole">
    <EuiButton>Only visible to users with enough privileges</EuiButton>
  </Authorize>
)
```